core/consensus/src/engine.rs

```diff
@@ -608,6 +641,19 @@ impl<Adapter: ConsensusAdapter + 'static> ConsensusEngine<Adapter> {
         // Save the block
         self.adapter.save_block(ctx.clone(), block.clone()).await?;
 
+        if let Ok(data) = HardforkInfoInner::decode(&block.header.extra_data) {
+            let mut self_proposal = self.node_info.hardfork_proposals.write().unwrap();
+
+            if self_proposal
+                .as_ref()
+                .map(|v| v == &data)
+                .unwrap_or_default()
+            {
+                // remove self hardfork proposal
+                self_proposal.take();
+            }
+        }
+
```

Comment:
Whether to remove a `hardfork_info` from `node_info` should be determined by the `hardfork info` in metadata.

Reply:
The generated block already contains hardfork information, then the proposal has been confirmed. At this time, the proposal for the next block does not need to be submitted for hardfork again.

Reply:
What I mean is that it is not rigorous to determine that hardfork info can be removed at the time of block generated, it should be after the info written to the metadata store. So the operation should be judgement by compare the hardfork info in metadata:

```rust
/// These code does not test
let store_hardfork_info = self.adapter.get_hardfork_info(); let mut self_proposal = self.node_info.hardfork_proposals.write().unwrap();

if self_proposal == store_hardfork_info {
	self_proposal.take();
}
```